File: Coding-Patterns/Coding-Patterns-Python/04_dynamic_programming/word_break.py
# LeetCode 139 - Word Break
# Difficulty: Medium
# Pattern: 1D DP + set lookup
#
# Problem: Given a string s and a dictionary wordDict, return True if
# s can be segmented into space-separated words from the dictionary.
#
# Approach: dp[i] = True if s[:i] can be segmented.
# For each i, try all j < i: if dp[j] and s[j:i] in wordDict -> dp[i] = True.
#
# Time: O(n^2)  Space: O(n)

from typing import List

def word_break(s: str, word_dict: List[str]) -> bool:
    word_set = set(word_dict)
    n = len(s)
    dp = [False] * (n + 1)
    dp[0] = True  # empty string is always valid

    for i in range(1, n + 1):
        for j in range(i):
            if dp[j] and s[j:i] in word_set:
                dp[i] = True
                break

    return dp[n]


# ---- Tests ----
if __name__ == "__main__":
    assert word_break("leetcode", ["leet", "code"]) == True
    assert word_break("applepenapple", ["apple", "pen"]) == True
    assert word_break("catsandog", ["cats", "dog", "sand", "and", "cat"]) == False
    print("All tests passed.")

File: Coding-Patterns/Coding-Patterns-CSharp/CodingPatterns/04_dynamic_programming/WordBreak.cs
// LeetCode 139 - Word Break
// Difficulty: Medium
// Pattern: 1D DP + HashSet lookup
//
// Time: O(n^2)  Space: O(n)
// Original approach: for each i, try all j < i, check s.Substring(j, i-j) in HashSet
//                    → O(n^2) substrings created + hashed
// Optimized:         Trie of dictionary words; for each i, walk the trie forward
//                    → no substring allocations; early termination
//
// Time: O(n * L) where L = max word length (typically << n^2)
// Space: O(total chars in dictionary)

namespace CodingPatterns.DynamicProgramming;

public class WordBreak
{
    public bool CanBreak(string s, IList<string> wordDict)
    {
        var wordSet = new HashSet<string>(wordDict);
        int n = s.Length;
        var dp = new bool[n + 1];
        dp[0] = true;

        for (int i = 1; i <= n; i++)
            for (int j = 0; j < i; j++)
                if (dp[j] && wordSet.Contains(s.Substring(j, i - j)))
                { dp[i] = true; break; }

        return dp[n];
    }
    private class TrieNode
    {
        public TrieNode[] Children = new TrieNode[26];
        public bool IsWord;
    }

    public bool CanBreakTrie(string s, IList<string> wordDict)
    {
        // Build trie
        var root = new TrieNode();
        foreach (var word in wordDict)
        {
            var node = root;
            foreach (var c in word)
            {
                int idx = c - 'a';
                node.Children[idx] ??= new TrieNode();
                node = node.Children[idx];
            }
            node.IsWord = true;
        }

        int n = s.Length;
        var dp = new bool[n + 1];
        dp[0] = true;

        for (int i = 0; i < n; i++)
        {
            if (!dp[i]) continue;

            // Walk the trie starting at s[i]
            var node = root;
            for (int j = i; j < n; j++)
            {
                int idx = s[j] - 'a';
                if (node.Children[idx] == null) break;  // no word extends this prefix
                node = node.Children[idx];
                if (node.IsWord) dp[j + 1] = true;
            }
        }
        return dp[n];
    }

    public static void Run()
    {
        var sol = new WordBreak();
        Console.WriteLine(sol.CanBreak("leetcode", new[] { "leet", "code" }));         // True
        Console.WriteLine(sol.CanBreak("applepenapple", new[] { "apple", "pen" }));     // True
        Console.WriteLine(sol.CanBreak("catsandog",
            new[] { "cats", "dog", "sand", "and", "cat" }));                             // False
    }
}
